### PR TITLE
Docker updates

### PR DESCRIPTION
**Pull request by ghost**
This is a follow-on PR to https://github.com/vivo-project/VIVO/pull/3839 and can be handled two ways; review this PR in place of https://github.com/vivo-project/VIVO/pull/3839 or maintain commits of branched PR by not squash merging.

**[VIVO GitHub issue](https://github.com/vivo-project/VIVO/issues)**: https://github.com/vivo-project/VIVO/issues/3822 and https://github.com/vivo-project/VIVO/issues/3786

Also related to https://github.com/vivo-project/VIVO/issues/3698 as the docker deployment supports specifying the Tomcat webapp context path.

# What does this pull request do?

- Improves docker image
  - not running tomcat as root
  - affords specifying webapp context path
  - supports loading sample data on reset of home
  - provides a subset of runtime properties to be defined on startup via environment variables
- Affords substitution of context.xml variables
  - vitro/rootUserAddress
  - vitro/defaultNamespace  

# What's new?
Updates to the Dockerfile, docker-compose.yml, and start.sh script among other deployment related configuration files.

# How should this be tested?
See README.md

# Interested parties
@chenejac

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,20 +1,42 @@
...
-ENV JAVA_OPTS="${JAVA_OPTS} -Dtdb:fileMode=$TDB_FILE_MODE"
+ENV SOLR_URL=http://localhost:8983/solr/vivocore
+ENV SELF_ID_MATCHING_PROPERTY=http://vivo.mydomain.edu/ns
 
-RUN mkdir /usr/local/vivo
-RUN mkdir /usr/local/vivo/home
+ENV LOAD_SAMPLE_DATA=false
+ENV SAMPLE_DATA_REPO_URL=https://github.com/vivo-project/sample-data.git
+ENV SAMPLE_DATA_BRANCH=main
+ENV SAMPLE_DATA_DIRECTORY=openvivo
```
Comment (chenejac) [inline Dockerfile:18]: I suppose this means openvivo data might be loaded in some reasonable time when load_sample_data flag is turned on (true).

File: installer/webapp/src/main/webResources/WEB-INF/classes/log4j.properties [hunk 2]
```diff
@@ -24,7 +24,7 @@
 # debug.log4j.properties exists will be used instead, if it exists, but is not stored in Subversion.
 
 log4j.appender.AllAppender=org.apache.log4j.RollingFileAppender
-log4j.appender.AllAppender.File=${catalina.base}/logs/${app-name}.all.log
+log4j.appender.AllAppender.File= ${catalina.base}/logs/${app-name}.all.log
```
Comment (chenejac) [inline installer/webapp/src/main/webResources/WEB-INF/classes/log4j.properties:27]:
```suggestion
log4j.appender.AllAppender.File=${catalina.base}/logs/${app-name}.all.log
```

File: start.sh [hunk 3]
```diff
@@ -2,59 +2,107 @@
...
+  echo "Using existing $VIVO_HOME_DIR/config/applicationSetup.n3"
+fi
+
+cp -r /tmp/vivo.war /usr/local/tomcat/webapps/$TOMCAT_CONTEXT_PATH.war
+
+export JAVA_OPTS="${JAVA_OPTS} -Dvivo-dir=$VIVO_HOME_DIR -Droot-user-address=$ROOT_USER_ADDRESS -Ddefault-namespace=$DEFAULT_NAMESPACE -Dtdb:fileMode=$TDB_FILE_MODE"
+
+echo "Giving time for Solr to startup..."
+sleep 15
+echo "Starting Tomcat"
```
Comment (chenejac) [inline start.sh:106]: someone recently reported this issue, thanks for this. Maybe 15 should also be a parameter/variable with 15 as default value.

**Review by chenejac — DISMISSED** [review]
@wwelling please check my comments here, I suggest to review this PR instead of previous one (https://github.com/vivo-project/VIVO/pull/3839), and of course we need to review Vitro PR as well (https://github.com/vivo-project/Vitro/pull/370)

File: start.sh (hunk 3 above)
Comment (wwelling) [reply start.sh:106]: This technically shouldn't be here. It is a hack and only working around a race condition when deploying using docker-compose. I will update to provide the timeout in the docker-compose file entrypoint opposed to it being in the start script of the container which will always sleep before startup.

**Comment by wwelling** [comment]
Dragan Ivanovic
  [3 days ago](https://vivo-project.slack.com/archives/C04Q2E3REES/p1677869185270699?thread_ts=1677856117.386259&cid=C04Q2E3REES)
Only bin, rdf and config files are copied into VIVO_HOME, right? I suppose some changes might happened into config and rdf in customized VIVO instances. I see the next scenarios:
1. An institution is installing VIVO for the first time, there is only empty VIVO_HOME directory.
2. An institution is upgrading VIVO, but didn't customized VIVO_HOME, meaning didn't change any single file in rdf and config.
3. An institution is upgrading VIVO to some future VIVO which is not changing any file which should be copied to config or rdf subdirectories of VIVO_HOME, but institution customized its previous VIVO version by changing something in VIVO_HOME/config or VIVO_HOME/rdf.
4. An institution is upgrading VIVO to some future VIVO which is changing some files which should be copied to config or rdf subdirectories of VIVO_HOME (introducing new property in default.runtime.properties, adding new n3 file in rdf, etc.), and institution also customized its previous VIVO version by changing something in VIVO_HOME/config or VIVO_HOME/rdf.

**Comment by wwelling** [comment]
```
find /vivo/home -type f | grep -E "^/vivo/home/bin/|^/vivo/home/config/|^/vivo/home/rdf/" | xargs md5sum > /vivo/home/digest.md5
```

**Comment by wwelling** [comment]
Merge conflicts with https://github.com/vivo-project/VIVO/pull/3940.

**Pull request closed without merging**